### PR TITLE
fix: Fixed check for duplicate JSObject names in an application page

### LINKED ISSUE #13692: [Bug]-[400]:error js function already exists after deleting
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description

to quote the user:
"It seems that the JS objects cache the functions you create, and there's no way to clear that cache.  I got an error that one of my controller functions already exists, even though I deleted it a while ago.  When I tried to delete and recreate the controller, it started saying that all the functions already existed, and the entire page started glitching.  I was able to get around this by cloning the page and deleting the old one, but this bug seems like something we'd want to fix."

### Steps To Reproduce

I was not able to reproduce this issue. Just posting here as reported by the user.

### Public Sample App

_No response_

### Version

Cloud

### LINKED ISSUE #14755: [Bug]: 400 error is returned on a newly defined JS function name
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Description

A newly defined function displays a message stating object and function name combination already exists when it does not.

### Steps To Reproduce

1. Go to app at [this](https://release.app.appsmith.com/app/scratch-pad/automation-test-cases-62987963c43c3d0bd6572185/edit/jsObjects/62b3f8abdd56222d3f8039b0?editName=false) location and observe the toast message and the browser logs display error for a newly defined function

image

### Public Sample App

_No response_

### Version

Cloud

## Test Plan
- [ ] [[Test Case]: verify that there is no 400 error appear on defining a new JS function name](https://github.com/appsmithorg/TestSmith/issues/1925)

Front logo [Front conversations](https://app.frontapp.com/open/top_5rtvw)

### PR DESCRIPTION
**Pull request by nidhi-nair**
## Description

Server side was allowing JS collection names to be duplicated in an app that took Appsmith apps to an inconsistent state. This PR fixes the logic for these checks.

Reproduction:
1. Create a new app, and open this app in another tab as well
2. Create a new JSObject in tab 1
3. Try to create a new JSObject in tab2. An error is seen on the page
4. Refresh the page
5. The app shows two JSObjects on the same name, but one of them won't have functions in the run dropdown.

Fixes #13692 
Fixes #14755

## Type of change
- Bug fix (non-breaking change which fixes an issue)

## How Has This Been Tested?
- JUnit test
- Manually tested

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: app/server/appsmith-server/src/test/java/com/appsmith/server/services/LayoutActionServiceTest.java [hunk 1]
```diff
@@ -544,6 +541,24 @@ public void tableWidgetKeyEscape() {
...
 
+    @Test
+    @WithUserDetails(value = "api_user")
+    public void testIsNameAllowed_withRepeatedActionCollectionName_throwsError() {
+        Mockito.doReturn(Flux.empty()).when(newActionService).getUnpublishedActions(Mockito.any());
+
+        ActionCollectionDTO mockActionCollectionDTO = new ActionCollectionDTO();
+        mockActionCollectionDTO.setName("testCollection");
+
+        Mockito.when(actionCollectionService.getActionCollectionsByViewMode(Mockito.any(), Mockito.anyBoolean()))
```
Comment (sondermanish) [inline app/server/appsmith-server/src/test/java/com/appsmith/server/services/LayoutActionServiceTest.java:552]: just for my info: why not use doReturn here as well, when both are spybeans?

**Review by sondermanish — APPROVED** [review]
when using doReturn can same stub be used more than once?

File: app/server/appsmith-server/src/test/java/com/appsmith/server/services/LayoutActionServiceTest.java (hunk 1 above)
Comment (nidhi-nair) [reply app/server/appsmith-server/src/test/java/com/appsmith/server/services/LayoutActionServiceTest.java:552]: I could have, this is a mix of convention plus the fact that the `doReturn` way of mocking a method completely bypasses the stubbed method. I like the when/thenReturn way of doing things better though, because of the type checks and the fact that you have to be aware of all the calls being made inside the method in case of spies. It is easier to catch edge cases like that. Does that make sense?

**Comment by Rhitottam** [comment]
/perf-test ref=rhito/feat/compare-with-lt-standard-deviation-for-regression

Testing new fixes to perf-infra tests

**Comment by Rhitottam** [comment]
/perf-test ref=rhito/feat/compare-with-lt-standard-deviation-for-regression

**Comment by Rhitottam** [comment]
/perf-test ref=rhito/feat/compare-with-lt-standard-deviation-for-regression
